sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/InsecureDeserialization.cs

```diff
@@ -0,0 +1,283 @@
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class InsecureDeserialization : SonarDiagnosticAnalyzer
+    {
+        private const string DiagnosticId = "S5766";
+        private const string MessageFormat = "Make sure not performing data validation after deserialization is safe here.";
+
+        private static readonly DiagnosticDescriptor Rule =
+            DiagnosticDescriptorBuilder
+                .GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager)
+                .WithNotConfigurable();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(VisitClassDeclaration, SyntaxKind.ClassDeclaration);
+
+        private static void VisitClassDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var classDeclaration = (ClassDeclarationSyntax)context.Node;
+            if (!HasConstructorsWithParameters(classDeclaration))
+            {
+                // If there are no constructors, or if these don't have parameters, there is no validation done
+                // and the class is considered safe.
+                return;
+            }
+
+            var typeSymbol = context.SemanticModel.GetDeclaredSymbol(classDeclaration);
+            if (!HasSerializableAttribute(typeSymbol))
+            {
+                return;
+            }
+
+            var implementsISerializable = ImplementsISerializable(typeSymbol);
+            var implementsIDeserializationCallback = ImplementsIDeserializationCallback(typeSymbol);
+
+            var walker = new ConstructorDeclarationWalker(context.SemanticModel);
+            walker.SafeVisit(classDeclaration);
+
+            if (!implementsISerializable &&
+                !implementsIDeserializationCallback)
+            {
+                foreach (var ctorInfo in walker.GetConstructorsInfo().Where(info => info.HasConditionalConstructs))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, ctorInfo.Declaration.GetLocation()));
+                }
+            }
+
+            if (implementsISerializable &&
+                !walker.HasDeserializationCtorWithConditionalStatements())
+            {
+                foreach (var ctorInfo in walker.GetConstructorsInfo().Where(info => !info.IsDeserializationConstructor && info.HasConditionalConstructs))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, ctorInfo.Declaration.GetLocation()));
+                }
+            }
+
+            if (implementsIDeserializationCallback &&
+                !OnDeserializationHasConditions(classDeclaration, context.SemanticModel))
+            {
+                foreach (var ctorInfo in walker.GetConstructorsInfo())
```

Comment:
Shouldn't we also filter on `.Where(info => info.HasConditionalConstructs)` to raise issue ? Won't this cause FP otherwise ?

Reply:
I was so sure that I've added that and I've completely forgotten 😄